radix-transaction-scenarios/src/scenarios/all_scenarios.rs

```diff
@@ -1,54 +1,422 @@
-            2 => Some(Box::new(|core| {
-                radiswap::RadiswapScenarioCreator::create(core)
-            })),
-            3 => Some(Box::new(|core| metadata::MetadataScenario::create(core))),
-            4 => Some(Box::new(|core| {
-                fungible_resource::FungibleResourceScenarioCreator::create(core)
-            })),
-            5 => Some(Box::new(|core| {
-                non_fungible_resource::NonFungibleResourceScenarioCreator::create(core)
-            })),
-            6 => {
-                Some(Box::new(|core| {
-                    account_authorized_depositors::AccountAuthorizedDepositorsScenarioCreator::create(core)
-                }))
+use account_authorized_depositors::AccountAuthorizedDepositorsScenarioCreator;
+use account_locker::AccountLockerScenarioCreator;
+use fungible_resource::FungibleResourceScenarioCreator;
+use global_n_owned::GlobalNOwnedScenarioCreator;
+use kv_store_with_remote_type::KVStoreScenarioCreator;
+use max_transaction::MaxTransactionScenarioCreator;
+use metadata::MetadataScenario;
+use non_fungible_resource::NonFungibleResourceScenarioCreator;
+use non_fungible_resource_with_remote_type::NonFungibleResourceWithRemoteTypeScenarioCreator;
+use radiswap::RadiswapScenarioCreator;
+use royalties::RoyaltiesScenarioCreator;
+use transfer_xrd::TransferXrdScenarioCreator;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ScenarioRequirements {
+    /// These scenarios do not require any protocol updates and are expected to be run at genesis
+    /// and be compatible with it.
+    Genesis,
+    /// These scenarios require one or more protocol updates and this enum specifies the protocol
+    /// update that they require.
+    ProtocolUpdateUpTo(ProtocolUpdate),
+}
+
+impl ScenarioRequirements {
+    pub fn all() -> &'static [ScenarioRequirements] {
+        &[
+            ScenarioRequirements::Genesis,
+            ScenarioRequirements::ProtocolUpdateUpTo(ProtocolUpdate::Anemone),
+            ScenarioRequirements::ProtocolUpdateUpTo(ProtocolUpdate::Bottlenose),
+        ]
+    }
+}
+
+macro_rules! define_scenario_builders {
+    (
+        $(
+            $requirement: expr => [
+                $($func: expr),* $(,)?
+            ]
+        ),* $(,)?
+    ) => {
+        pub fn scenario_builders() -> IndexMap<
+            ScenarioRequirements,
+            Vec<Box<dyn FnOnce(ScenarioCore) -> Box<dyn ScenarioInstance>>>
+        > {
+            ::sbor::prelude::indexmap! {
+                $(
+                    $requirement => vec![
+                        $(
+                            Box::new($func)
+                                as Box<dyn FnOnce(ScenarioCore) -> Box<dyn ScenarioInstance>>
+                        ),*
+                    ]
+                ),*
+            }
+        }
+    };
+}
+
+define_scenario_builders! {
+    // The set of scenarios that should be run after genesis.
+    ScenarioRequirements::Genesis => [
```

Comment:
In my opinion, the requirements of the scenario should live in the scenario itself. That way, if I want to execute scenario X, I can check if it's relevant from the scenario itself, rather than it living out here.

We can do this in the node, for example.

Reply:
(This can be on `ScenarioMetadata`)